ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -0,0 +1,560 @@
+  }
+  if (type_mask & ProfileType::Wall) {
+    val_idx.wall_time = get_value_idx("wall-time", "nanoseconds");
+    val_idx.wall_count = get_value_idx("wall-samples", "count");
+  }
+  if (type_mask & ProfileType::Exception) {
+    val_idx.exception_count = get_value_idx("exception-samples", "count");
+  }
+  if (type_mask & ProfileType::LockAcquire) {
+    val_idx.lock_acquire_time =
+        get_value_idx("lock-acquire-wait", "nanoseconds");
+    val_idx.lock_acquire_count = get_value_idx("lock-acquire", "count");
+  }
+  if (type_mask & ProfileType::LockRelease) {
+    val_idx.lock_release_time =
+        get_value_idx("lock-release-hold", "nanoseconds");
+    val_idx.lock_release_count = get_value_idx("lock-release", "count");
+  }
+  if (type_mask & ProfileType::Allocation) {
+    val_idx.alloc_space = get_value_idx("alloc-space", "bytes");
+    val_idx.alloc_count = get_value_idx("alloc-samples", "count");
+  }
+  if (type_mask & ProfileType::Heap) {
+    val_idx.heap_space = get_value_idx("heap-space", "bytes");
+  }
+
+  values.resize(samplers.size());
+  std::fill(values.begin(), values.end(), 0);
+
+  ddog_prof_Period default_sampler = {
+      samplers[0], 1}; // Mandated by pprof, but probably unused
+  ddog_profile = ddog_prof_Profile_new({&samplers[0], samplers.size()},
+                                       &default_sampler, nullptr);
+
+  // Prepare for use
+  reset();
+
+  // Initialize the size for buffers
+  locations.reserve(2040);
+  lines.reserve(2048);
+  strings.reserve(8192);
+}
+
+Profile::~Profile() { ddog_prof_Profile_drop(ddog_profile); }
+
+bool Profile::reset() {
+  if (!ddog_prof_Profile_reset(ddog_profile, nullptr)) {
+    errmsg = "Unable to reset profile";
+    return false;
+  }
+  return true;
+}
+
+bool Profile::start_sample(unsigned int nframes) {
+  strings.clear();
+  clear_buffers();
+  this->nframes = nframes;
+  return true;
+}
+
+void Profile::push_frame_impl(std::string_view name, std::string_view filename,
+                              uint64_t address, int64_t line) {
+
+  // Ensure strings are stored.
+  // Slightly wasteful since it requires allocating another string.
+  auto insert_or_get = [&](std::string_view sv) -> std::string_view {
+    std::string str(sv);
+    auto [it, _] = strings.insert(std::move(str));
+    return *it;
+  };
+  name = insert_or_get(name);
+  filename = insert_or_get(filename);
+
+  lines.push_back({
+      .function =
+          {
+              .name = to_slice(name),
+              .system_name = {},
+              .filename = to_slice(filename),
+              .start_line = 0,
+          },
+      .line = line,
+  });
+
+  locations.push_back({
+      {},
+      address,
+      {&lines.back(), 1},
+      false,
+  });
+}
+
+void Profile::push_frame(std::string_view name, std::string_view filename,
+                         uint64_t address, int64_t line) {
+
+  push_frame_impl(name, filename, address, line);
```

Comment:
The python profiler has a user-configurable maximum stack depth, and when it exceeds that depth it appends a single "omitted frames" frame to the top of the stack (err... bottom of the flamegraph, err... icicle chart).  I used to implement this on the Python side, but it required putting the same logic in multiple places.  Instead, I consolidated it in the native code since your last review.

That check (at the top of flush_sample) reuses the same `push_frame_impl()` logic, but doesn't check for the `max_nframes`.

Ah; I see I elided the check in `push_frames()` in this version.  :)

This is actually enforced by the Python side, but I wanted to be more strict.  I fixed the missing check. Oops!